Ak #1290 view visibility dropinui

abhishek1508 opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->
Fixes https://github.com/mapbox/navigation-sdks/issues/1290

<!--
---------- CHECKLIST ----------
1. Add related labels (`bug`, `feature`, `new API(s)`, `SEMVER-MAJOR`, `needs-backporting`, etc.).
2. Update progress status on the project board.
3. Request a review from the team, if not a draft.
4. Add targeted milestone, when applicable.
5. Create ticket tracking addition of public documentation pages entry, when applicable.
-->

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/state/ManeuverState.kt:
```diff
@@ -0,0 +1,29 @@
+                isVisible = false,
+                navigationState = NavigationState.Empty,
+                maneuver = ExpectedFactory.createError(
+                    ManeuverErrorFactory.buildManeuverError(
```
I had to create a new `ManeuverErrorFactory` that would allow to construct this object. Since we have marked `*Factory` with @ExperimentalMapboxNavigationAPI, this has taken affect upstream to `MainActivity`. We need to make a decision if it's ok to add this tag to an Activity, or do we want to remove this tag from any `*Factory` objects that we create.
/cc @LukasPaczos

LukasPaczos replied:
@abhishek1508 we don't have to mark all classes that use the new experimental factory with `@ExperimentalMapboxNavigationAPI` (which then goes up and up the usage tree), we just need to tag the specific internal usages with `@OptIn(ExperimentalMapboxNavigationAPI::class)` which does not bubble up.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ManeuversContract.kt:
```diff
@@ -0,0 +1,39 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+
+sealed class ManeuversAction {
```
This shouldn't be an action that developers have access to. This is an internal logic (and method of transferring data), so this and all other classes below should all be internal.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -143,46 +180,141 @@ class NavigationView : ConstraintLayout, LifecycleObserver {
-                renderVisibility(state)
+            speedLimitViewModel.speedLimitState().collect { state ->
+                binding.speedLimitContainer.visibility = state.isVisible.toVisibility()
+                (speedLimitView as MapboxSpeedLimitView).render(state.speedLimit)
```
We'll need to have a more extensive logic here and only run an update if the view actually is an instance of `MapboxSpeedLimitView` (same for all other views). If someone would use a custom view for `speedLimitView`, we'd crash here.

If it's not our view, we shouldn't even subscribe to the view model for any updates. We need a controlling unit that understands the configuration and which views are ours and which aren't so that we can skip them.

LukasPaczos replied:
Let's also try to move it out of the `NavigationView` class so that it doesn't grow.

abhishek1508 replied:
> If it's not our view, we shouldn't even subscribe to the view model for any updates. We need a controlling unit that understands the configuration and which views are ours and which aren't so that we can skip them.

I agree with this totally and was already planning to cut a ticket on this task. I will remove this line here in this PR as this might need some thoughts on how to do even though I have something in mind. But, let's not block this PR and introduce a proper system for it separately.

> Let's also try to move it out of the NavigationView class so that it doesn't grow.

Do you mean all `*.collect { }` calls as there are several

/cc @LukasPaczos

cafesilencio commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -48,6 +67,19 @@ class NavigationView : ConstraintLayout, LifecycleObserver {
             )
         }
     }
+    private val distanceFormatter = MapboxDistanceFormatter(
```
I don't think the NavigationView should be creating all of these dependencies for other classes like view models. Just because we're not using a dependency injection framework doesn't mean the NavigationView should take on this role. The providing of dependencies should be abstracted from the view models as well as the NavigationView. No need to hold up this PR for this issue. I'll propose an alternative to this separately.

cafesilencio replied:
In fact the view model factories would be a good place to move these to.

abhishek1508 replied:
Yep I agree and I had the same intention to improve this later and focus on the item relevant to this PR. Thanks!

cafesilencio commented:
PR https://github.com/mapbox/mapbox-navigation-android/pull/5086 is landing momentarily. After it's merged please make adjustments to `NavigationView` so that rather than public functions, the calls are exposed through the `MapboxNavigationViewApi`.

cafesilencio commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -112,13 +137,15 @@ class NavigationView : ConstraintLayout {
     private val cameraViewModel: CameraViewModel by lazy {
         ViewModelProvider(activity)[CameraViewModel::class.java]
     }
+    private lateinit var volumeViewModel: SoundButtonViewModel
```
These should be immutable IMO.

abhishek1508 commented:
Hey @LukasPaczos 

Based on our discussion, I have updated the code in this PR to reflect 2 changes:
- Viewmodels are now responsible to read the last value using the flow.
- I have tried to move the `.*collect` calls out of the `NavigationView` as it is growing rapidly. For that effect, instead of arranging the package structure horizontally (contract, viewmodel, state, processor), I have now arranged the package structure vertically based on features(maneuver, recenter, routeoverview, sound, tripprogress, speedlimit). The concept is that DropIn is made up of several standalone components and so I have introduced a `.*Component` class for every feature. For ex:
```
class ManeuverComponent(
    val view: View,
    val container: FrameLayout,
    val viewModel: ManeuverViewModel
)
```
This has allowed me to trigger all actions in the `*.Component` as well as hook up all the state observer changes in the components. One important place where I want a second thought is would this `maneuverComponent` be garbage collected when `NavigationView` is destroyed on configuration change.

abhishek1508 commented:
@LukasPaczos This is ready for another round of review. Given the time constraint and the fact that this PR is growing fast and needs to merge asap I have created this [ticket](https://github.com/mapbox/navigation-sdks/issues/1316) to track adding the missing unit tests. I would also not worry about the `internalObservers` I have added at the moment. We can improve it in further iterations.

LukasPaczos commented on gradle/jacoco.gradle:
```diff
@@ -7,4 +7,5 @@ jacoco {
 tasks.withType(Test) {
   // needed to capture coverage from Robolectric tests
   jacoco.includeNoLocationClasses = true
+  jacoco.excludes = ['jdk.internal.*']
```
Why is this needed here? I imagine you were running into some problems with unit tests and Java 11?

abhishek1508 replied:
Yeah let me try a different Java version and see if that fixes the problem

LukasPaczos replied:
Vitaly mentioned that he'll help to drive https://github.com/mapbox/mapbox-navigation-android/pull/4454 to the finish line so let's remove this change from here for now.

LukasPaczos commented on examples/src/main/java/com/mapbox/navigation/examples/core/MapboxDropInActivity.kt:
```diff
@@ -4,10 +4,13 @@ import android.os.Bundle
 
     private lateinit var binding: LayoutActivityDropInBinding
+
+    @ExperimentalCoroutinesApi
```
This can be removed.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapboxNavigationViewApi.kt:
```diff
@@ -29,73 +35,126 @@ interface MapboxNavigationViewApi {
     fun configureNavigationView(viewProvider: ViewProvider)
 }
 
+@ExperimentalCoroutinesApi
```
This can be removed or changed to `@OptIn(ExperimentalCoroutinesApi::class)`.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -26,38 +26,56 @@ import com.mapbox.maps.plugin.gestures.OnMapLongClickListener
-import java.lang.IllegalStateException
 import java.lang.ref.WeakReference
 import java.util.concurrent.CopyOnWriteArraySet
 
```
The `@ExperimentalCoroutinesApi` should be changed to `@OptIn(ExperimentalCoroutinesApi::class)`.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -80,33 +98,73 @@ class NavigationView : ConstraintLayout {
-    private val mapInitOptions: MapInitOptions
-    private val navigationLocationProvider = NavigationLocationProvider()
+    private val onPositionChangedListener = OnIndicatorPositionChangedListener { point ->
+        mapView.getMapboxMap().getStyle()?.let { style ->
+            routeLineViewModel.positionChanged(point, style)
+        }
```
Why do we not have a `MapboxRouteLineComponent`? Not a blocker, just wondering if this is intentional or planned as follow up.

abhishek1508 replied:
The route line just got merged yesterday, so probably was missed.

LukasPaczos replied:
Okay, feel free to add here or cut a ticket :+1:

abhishek1508 replied:
Added `RouteLineUIComponent` and `RouteArrowUIComponent`
@LukasPaczos

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -298,145 +424,169 @@ class NavigationView : ConstraintLayout {
+                    it.onRouteProgressChanged(routeProgress)
+                }
+                uiComponents.forEach { uiComponent ->
+                    when (uiComponent) {
+                        is ManeuverUIComponent.MapboxManeuverUIComponent -> {
+                            uiComponent.onRouteProgressChanged(routeProgress)
+                        }
+                        is TripProgressUIComponent.MapboxTripProgressUIComponent -> {
+                            uiComponent.onRouteProgressChanged(routeProgress)
+                        }
```
```suggestion
                    when (uiComponent) {
                        is RouteProgressObserver -> {
                            uiComponent.onRouteProgressChanged(routeProgress)
                        }
```

Same below where appropriate.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -298,145 +424,169 @@ class NavigationView : ConstraintLayout {
+                        }
                     }
                 }
+                navigationLocationProvider.changePosition(
+                    locationMatcherResult.enhancedLocation,
+                    locationMatcherResult.keyPoints
+                )
```
We should have a `component` to manage the puck and wire these operations through that `component`.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:
```diff
@@ -298,145 +424,169 @@ class NavigationView : ConstraintLayout {
+                externalLocationObservers.forEach {
+                    it.onNewRawLocation(locationUpdate)
+                }
+                cameraViewModel.consumeLocationUpdate(locationUpdate)
```
Same as above, we should have a component for the camera. Can be added separately though.

abhishek1508 replied:
I agree and I think it would make sense to address this change and above in your camera PR.

LukasPaczos replied:
:ok_hand:

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:
```diff
@@ -0,0 +1,94 @@
+                    val processor = ManeuverProcessor.ProcessNavigationState(value.navigationState)
+                    processor.process()
+                }
+                is ManeuverAction.UpdateVisibility -> {
+                    val processor = ManeuverProcessor.ProcessVisibility(
+                        _maneuverState.value.navigationState
+                    )
+                    processor.process()
```
We don't need a distinct "update visibility" action. Visibility is based on state and data which are both separate actions.

LukasPaczos replied:
Same everywhere below.

abhishek1508 replied:
I had it exactly like that first where a change in `NavigationState` would also update the visibility. Although it makes complete sense, I think we loose some control here and out of nowhere without any intention changing the state changes the visibility as well. I agree it is probably verbose, but it makes more sense to me trigger the action to update visibility rather than some other action that updates a different part of the state. If we would have to travel down the history and track actions, it would become difficult to figure out which action changed the visibility. Also, if in the future we were to change the fact that visibility is not dictated by navigation state then not having a separate action for it will break.

LukasPaczos replied:
Why would a change in state ever not change the visibility, or how would we change the state unintentionally? It seems redundant to me that we need to call a visibility change action, it's more of a source of bugs if we _forget_ to call it after the state change. Following the example, we'd have to have an empty action for each operation (maneuver update, distance remaining update, etc).

I still think that we can remove `UpdateVisibility` action and only reduce the available data into a result which is applied, without empty triggers like this.

> Also, if in the future we were to change the fact that visibility is not dictated by navigation state then not having a separate action for it will break.

That would be a breaking change for the entire `component` anyway so that might also be a better point in time to add the necessary action instead of introducing it now without a clear use-case.

abhishek1508 replied:
> it's more of a source of bugs if we forget to call it after the state change. Following the example, we'd have to have an empty action for each operation (maneuver update, distance remaining update, etc).

Agreed with this. We could probably remove the empty `UpdateVisibility`, but we would later have to add a new action UpdateVisibility(show = true/false) if an end user wants to show a view which is hidden or vice versa
cc @LukasPaczos

LukasPaczos replied:
> we would later have to add a new action UpdateVisibility(show = true/false) if an end user wants to show a view which is hidden or vice versa

I still hope this will be controlled by the state variable which would have an understanding of which component should be visible rather than components hardcoding if they are visible for given state (as things currently are).

abhishek1508 replied:
That would be an ideal scenario. In reality the visibility of a view is controlled by many different factors here and is not the same for every view. `Recenter` uses `navigationState` and `cameraState` whereas `Maneuver` uses `navigationState` and `maneuverValue`. It is probably too soon to say if there are any other states/data point that dictate the visibility. Also, it's probably difficult to predict if an end user would just want to show the view irrespective of how it is dictated by the state/data. 

However, we can remove it for now and cross that bridge when we come to it.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:
```diff
@@ -0,0 +1,30 @@
+import com.mapbox.navigation.ui.maneuver.model.ManeuverErrorFactory
+
+data class ManeuverState(
+    val isVisible: Boolean,
+    val navigationState: NavigationState,
+    val maneuver: Expected<ManeuverError, List<Maneuver>>
```
Here only `isVisible` and `maneuver` are actual result values which are then used to update the UI. `navigationState` is a data point out of which the result is computed.

We might want to separate these to have 2 object, one object that holds `isVisible` and `maneuver` state, and the other that holds previous input values for reprocessing when values are updated.

LukasPaczos replied:
Same in all other components.

abhishek1508 replied:
I am wondering what is the advantage of separating then out into separate components?

LukasPaczos replied:
Readability in my opinion. It wasn't immediately clear to me why do we store the `navigationState` here since it is not used for any view operations directly.

I understand that it's here to be able to reduce the state when new route progress comes in but it might be worth decoupling the actual state that views feed off of from the internal view model's state.

LukasPaczos replied:
Not a very strong point here though, just my small observation. We can address this later as we add more state properties and see what fits where.

abhishek1508 replied:
Yep. I believe we are still not clear on what other data points would be required for our components. It would probably make more sense to update this if required once we have a good amount of functionality ready so then we have more clear understanding. For ex: we are at the moment not tracking any UI state like storing the recycler view scroll position that would help to show the recycler view from the position you left it.

LukasPaczos reviewed (approved):
Great work!

Let's ship and follow up on the remaining changes that we've been tracking separately.
